### ISSUE TITLE
Allow certain hook to be verbose

### ISSUE THREAD

**Issue by bagerard** [pre-commit/pre-commit#689]
Hi,
While integrating flake8 as a hook, I'm trying to have certain checks (like complexity warnings) run on every commit but I don't necessarily want them to be blocking. flake8 has this very nice --exit-zero flag that allows this but the only way to get the output printed through pre-commit is to run pre-commit with --verbose but this prints a ton of useless information from other hooks.

My suggestion is to allow a 'verbose' flag to be set per hook as opposed to existing --verbose which is set globally. IMO it would allow to have extremely valuable information out of flake8 and pylint for instance. I could even run part of flake8 as being blocking, and other part as being non-blocking.
Let me know your thoughts, I'd be happy to work on this

**Comment by asottile** [comment]
First off, thanks for the issue :)

So I thought I had addressed this somewhere else before but I couldn't find it -- it's possible I'm remembering a private discussion about the same thing.  I'll try and reiterate my thoughts on that here and we can go from there!

One design goal of a lot of unix utilities is "silence is golden" and for pre-commit the meaning takes a bit of a different form.  As pre-commit is intended to be ~somewhat of a gateway for vetting changes it should generally be as quiet as possible unless something goes wrong.  In other words, attempt to keep as high of a signal-to-noise ratio as possible.  This is also suggested in the design of hooks that plug into the framework ([an example](https://github.com/tox-dev/tox/pull/708#issuecomment-350567303)) -- by default they provide just as much information to answer why they're stopping the commit at all (preferring "filename.py: import reordered" over a full-on diff of a file).

The previous ~similar systems that `pre-commit` replaced had a huge verbosity problem in that _most_ of their output was "warnings".  Yes the messages were probably useful if you were going to take the time to read them, but for 99% of developers if it wasn't stopping them from committing they did not care and ignored the messages.  Not only did this decrease the likelihood that these developers would fix the problems presented in warnings, but it also hurt the attentiveness to messages in other (related **_and_** unrelated) commandline tools.  I'm probably stealing the name from somewhere but I usually refer to this problem as _warning fatigue_.  More succinctly, if users are presented with lots of non-critical information, they are trained to ignore all of the (important) output.

I think you'll find that while you may read the warning messages about code complexity and such, other contributors will filter this out -- and worse may filter out _all_ of pre-commit's messaging.

All that said, implementing such a thing is trivial (though I personally wouldn't use it and I would generally discourage anyone from using it ;) ).

[here](https://github.com/pre-commit/pre-commit/blob/98ec74dcabd8279f99b072e4232c5042ce16924b/pre_commit/commands/run.py#L133)

```diff
-    if (stdout or stderr or file_modifications) and (retcode or args.verbose):
+    if (
+            (stdout or stderr or file_modifications) and
+            (retcode or args.verbose or hook['verbose'])
+    ):
```

(plus the changes to validating + defaulting this in `pre_commit/clientlib.py`, plus a quick regression test in tests/commands/run_test.py)

Let me know what you think!

**Comment by asottile** [comment]
@bagerard curious to hear your thoughts on this -- if you can come up with a strong use case I'd be happy to entertain a pull request to add this :)

**Comment by bagerard** [comment]
Sorry for the late reply and thanks for the detailed feedback.
I understand your point of view and feel the same about 'warning fatigue'. Last thing I want is that my team ignores messages or start bypassing pre-commit's. I still personally believe that if it is used sparingly/smartly, it could be valuable.

The use case I had in mind was having like a very short list of flake8's checks enabled like:
- flake8's complexity check
- E721 do not compare types, use ‘isinstance()’
- E711 comparison to None should be ‘if cond is None

Those are checks that are optional and that could point out false positives. Knowing our codebase, those checks would only rarely print warnings. If it would start printing things all the time, we'd re-consider it indeed.

I'll discuss it with my team members and get back to you shortly.

**Comment by bagerard** [comment]
@asottile I discussed with my team members and they had the feeling that it could be valuable as well if used sparingly. May I ago ahead and work on a PR?

**Comment by asottile** [comment]
Sure, let's see what this looks like in PR form

**Comment by asottile** [comment]
via #695

**Comment by asottile** [comment]
This has been released as part of [v1.6.0](https://github.com/pre-commit/pre-commit/releases/tag/v1.6.0)!

**Comment by bagerard** [comment]
![image](https://user-images.githubusercontent.com/11854690/35782133-c0107e1a-09f3-11e8-83d7-a21e963e67b6.png)
Thanks!
